### PR TITLE
ppx: consistent errors in runtime

### PR DESCRIPTION
**Pull request by andreypopp**
The main thing we fix here is the consistent use of exceptions in native runtime, which used, before this commit either its own Of_json_error or yojson's Type_error.

Then there's
```
val of_string : string -> t
```
which is not required by ppx but was introduced so that the runtime modules can function as minimal JSON interfaces usable both in native and browser environments (so called universal workflow).

Before this commit the error behaviour was not specified, now we introduce `exception Of_json of string` in both native and browser and make `of_string` raise it in case of invalid JSON.

Ref: #9

### CODE REVIEW

File: ppx/browser/ppx_deriving_json_runtime.ml [hunk 1]
```diff
@@ -3,8 +3,23 @@ type t = Js.Json.t
 let to_json t = t
 let of_json t = t
 let to_string t = Js.Json.stringify t
-let of_string s = Js.Json.parseExn s
-let of_json_error msg = raise @@ Json.Decode.DecodeError msg
+
+exception Json_error of string
```
Comment (davesnx) [inline ppx/browser/ppx_deriving_json_runtime.ml:7]: encode_error?

File: ppx/browser/ppx_deriving_json_runtime.ml [hunk 2]
```diff
@@ -3,8 +3,23 @@ type t = Js.Json.t
...
+
+let of_string s =
+  try Js.Json.parseExn s
+  with exn ->
+    let msg =
+      match Js.Exn.asJsExn exn with
+      | Some jsexn -> Js.Exn.message jsexn
+      | None -> None
+    in
+    let msg = Option.value msg ~default:"JSON error" in
```
Comment (davesnx) [inline ppx/browser/ppx_deriving_json_runtime.ml:17]: I would add a comment here, explaining that Js.Exn.message can't be none in the browser (and probably true in any js environment)

**Review by davesnx — APPROVED** [review]
Looks very good

File: ppx/browser/ppx_deriving_json_runtime.ml (hunk 1 above)
Comment (andreypopp) [reply ppx/browser/ppx_deriving_json_runtime.ml:7]: no, this is JSON parsing error, not encode

File: ppx/browser/ppx_deriving_json_runtime.ml (hunk 2 above)
Comment (andreypopp) [reply ppx/browser/ppx_deriving_json_runtime.ml]: done

File: ppx/browser/ppx_deriving_json_runtime.ml (hunk 1 above)
Comment (davesnx) [reply ppx/browser/ppx_deriving_json_runtime.ml:7]: Parsing_error then? ^^

File: ppx/browser/ppx_deriving_json_runtime.ml [hunk 3]
```diff
@@ -3,8 +3,26 @@ type t = Js.Json.t
...
+let of_string s =
+  try Js.Json.parseExn s
+  with exn ->
+    let msg =
+      match Js.Exn.asJsExn exn with
+      | Some jsexn -> Js.Exn.message jsexn
+      | None -> None
+    in
+    let msg =
+      (* msg really cannot be None in browser or any sane JS runtime *)
```
Comment (jchavarri) [inline ppx/browser/ppx_deriving_json_runtime.ml:18]: The reason is because the exn raised by `Js.Json.parseExn` is a JavaScript one, right?

File: ppx/browser/ppx_deriving_json_runtime.ml (hunk 1 above)
Comment (jchavarri) [reply ppx/browser/ppx_deriving_json_runtime.ml:7]: Or maybe `Of_string_error` or `String_to_json_error`. I find a bit confusing that this is called `Json_error` and the errors happening at the later stage are called `Of_json_error`.

File: ppx/native/ppx_deriving_json_runtime.ml [hunk 4]
```diff
@@ -28,19 +47,43 @@ module To_json = struct
...
+    | json -> of_json_error_type_mismatch json "float"
 
   let unit_of_json = function
     | `Null -> ()
     | _ -> of_json_error "expected null"
 
-  let option_of_json v_of_json = Yojson.Basic.Util.to_option v_of_json
+  let option_of_json v_of_json = function
+    | `Null -> None
+    | json -> Some (v_of_json json)
```
Comment (jchavarri) [inline ppx/native/ppx_deriving_json_runtime.ml:82]: Why was this change needed? There are no errors happening.

File: ppx/native/ppx_deriving_json_runtime.ml [hunk 5]
```diff
@@ -3,12 +3,31 @@ type t = Yojson.Basic.t
...
+
+let of_string s =
+  try Yojson.Basic.from_string s
+  with Yojson.Json_error msg -> raise (Json_error msg)
 
 exception Of_json_error of string
 
 let of_json_error msg = raise (Of_json_error msg)
 
+let show_js_type = function
```
Comment (jchavarri) [inline ppx/native/ppx_deriving_json_runtime.ml:17]: In this case "js" is an abbreviation of "json" right? As there's a js runtime it's a bit confusing.

**Review by jchavarri — APPROVED** [review]
Looks good. Just minor suggestions or questions.

File: ppx/browser/ppx_deriving_json_runtime.ml (hunk 3 above)
Comment (andreypopp) [reply ppx/browser/ppx_deriving_json_runtime.ml:18]: yes

File: ppx/native/ppx_deriving_json_runtime.ml (hunk 5 above)
Comment (andreypopp) [reply ppx/native/ppx_deriving_json_runtime.ml]: fixed to be `show_json_type`

File: ppx/browser/ppx_deriving_json_runtime.ml (hunk 1 above)
Comment (andreypopp) [reply ppx/browser/ppx_deriving_json_runtime.ml]: Fixed to be `Of_string_error`.
